cid/drainer.py

```diff
@@ -0,0 +1,168 @@
+
+        self.ecs_client.update_container_instances_state(
+            cluster=cluster, containerInstances=[container_instance_arn], status='DRAINING',
+        )
+
+        self.cache['InstanceIsDraining'] = True
+
+    def get_running_tasks(self, cluster, container_instance_arn):
+        """Return a list of tasks running on a container instance.
+           :param cluster: ECS cluster of which ECS container instance of interest is a member
+           :param container_instance_arn: ECS container instance ARN
+        """
+        logger = self.logger
+
+        list_tasks_response = self.ecs_client.list_tasks(
+                cluster=cluster, containerInstance=container_instance_arn)
+
+        logger.debug("Task list for container instance: %s: %s",
+                     container_instance_arn, list_tasks_response['taskArns'])
+
+        running_tasks = list_tasks_response['taskArns']
+
+        logger.debug("Container instance: %s has tasks: %s", container_instance_arn, running_tasks)
+
+        return running_tasks
+
+    def run(self):
+        logger = self.logger
+        message = self.message
+
+        logger.info("Lambda received the event %s", self.event)
+
+        asg_name = message['AutoScalingGroupName']
+        topic_arn = self.event['Records'][0]['Sns']['TopicArn']
+
+        if not ('LifecycleTransition' in message and
+                message['LifecycleTransition'] == 'autoscaling:EC2_INSTANCE_TERMINATING'):
+            logger.info("Ignoring non-EC2_INSTANCE_TERMINATING message")
+            return
+
+        ec2_instance_id = message['EC2InstanceId']
+        logger.info("Received termination notification for EC2 instance-id: %s, ASG: %s",
+                    ec2_instance_id, asg_name)
+
+        cluster, container_instance_arn = self.get_ecs_details(ec2_instance_id)
+        logger.info(
+            "EC2 instance-id: %s appears to be container instance ARN: %s, belonging to ECS cluster: %s",
+            ec2_instance_id, container_instance_arn, cluster
+        )
+
+        lifecycle_hook_name = message['LifecycleHookName']
+        logger.debug("Lifecycle hook name is %s", lifecycle_hook_name)
+
+        self.set_draining(cluster, container_instance_arn)
+
+        logger.info("Checking for running tasks...")
+        tasks_running = self.get_running_tasks(cluster, container_instance_arn)
+        logger.debug("Running tasks: %s", tasks_running)
+
+        if len(tasks_running) > 0:
+
+            logger.info("There are still running tasks on %s, invoking myself again to re-check",
+                        container_instance_arn)
+
+            cache = self.cache
+            cache['EcsCluster'], cache['ContainerInstanceArn'] = cluster, container_instance_arn
+            message['_CidLambdaCache'] = cache
+
+            self.logger.info("Publishing to SNS topic: %s", topic_arn)
+            self.sns_client.publish(TopicArn=topic_arn, Message=json.dumps(message),
```

Comment:
is there any delay before the next invocation?

Reply:
No. In that regard I'm just copying the AWS code. But sleeping in a Lambda doesn't make sense unless your invocation is a tiny fraction of a second.

Reply:
Well, I think that instance draining will take quite some time. So invoking once per second or once per 3 seconds will be better. Now it seems like it will be running quite often.